Add AutoIt syntax highlighting and Ctags parser for AutoIt

Skif-off opened the pull request:
#1017 continuation, new (last) attempt.

b4n commented on scintilla/scintilla_changes.patch:
```diff
@@ -74,7 +74,7 @@ index ed47aa8..e58f1ab 100644
 -	LINK_LEXER(lmAsn1);
 -	LINK_LEXER(lmASY);
--	LINK_LEXER(lmAU3);
+ 	LINK_LEXER(lmAU3);
```
You also need to change the diff section header to match (`-77,123 +77,51`) as you're changing the final line count.

Skif-off replied:
I will check and fix it.

elextr commented:
To be clear, b4n's changes on the parser are for ctags and he should have made them there if he doesn't like the parser :)

Do not make the parser in Geany different to upstream, we have enough of those already!!!

b4n commented:
@elextr yeah, I should have made them on the PR integrating the parser to UCTags, but I missed that train. I'm commenting here because it's an opportunity, but yes any change should be made in UCTags, and then imported.

Skif-off commented on data/filedefs/filetypes.autoit:
```diff
@@ -0,0 +1,72 @@
+commentblock=comment
+number=number
+function=function
+keyword=type
```
Hm, I don't remember why I used this style, I will fix it.

Skif-off commented on data/filedefs/filetypes.autoit:
```diff
@@ -0,0 +1,72 @@
+[styling]
+# Edit these in the colorscheme .conf file instead
+default=default
+comment=comment
```
Usually AutoIt's line and block comments are highlighted equally and I did it this way. Is it mistake?

All ```filetypes*``` are editable files, if needed :)

b4n replied:
We have 2 named styles: one for single-line comments, and one for multi-line comments.  While most of the colorschemes (if not all) use the same colors for both, it's a good thing to still differentiate those in the filetypes files to so a colorscheme that would want to style them different would work consistently with all filetypes.

Skif-off replied:
Ok, I will fix it.

Skif-off commented on data/filedefs/filetypes.autoit:
```diff
@@ -0,0 +1,72 @@
+macro=preprocessor
+string=string
+operator=operator
+variable=keyword_2
```
```identifier == default``` in all Geany's colorshemes (default and from [here](https://github.com/geany/geany-themes)), but I wanted to use the highlighting.
Some```filetypes*``` has ```variable=type```, but I already use ```type``` and I took ```keyword_2``` because that it no one uses and it looks normal in all colorschemes.

I don't know what to do here.

b4n replied:
OK… I'm not sure either what to do here, and I guess this is one of the instances where the named styles we provide are not always enough.

The actual problem is possibly more subtle, it `identifier` might be styled as `default` by most colorschemes because doing otherwise results in weird behavior in some filetypes, which might be for a bad reason; or it might be that we don't have enough subtleties in the styles we provide.

@codebrainz do you have an opinion?

Skif-off commented on data/filedefs/filetypes.autoit:
```diff
@@ -0,0 +1,72 @@
+
+[keywords]
+keywords=and byref case const continuecase continueloop default dim do else elseif endfunc endif endselect endswitch endwith enum exit exitloop false for func global if in local next not null or redim return select static step switch then to true until volatile wend while with
+functions=abs acos adlibregister adlibunregister asc ascw asin assign atan autoitsetoption autoitwingettitle autoitwinsettitle beep binary binarylen binarymid binarytostring bitand bitnot bitor bitrotate bitshift bitxor blockinput break call cdtray ceiling chr chrw clipget clipput consoleread consolewrite consolewriteerror controlclick controlcommand controldisable controlenable controlfocus cont ...
```
Sorry, I don't understand :) Lists of keywords and internal functions are normal practice (Lisp, Latex, Fortran, FreeBasic, Python, Lua and so on).

Skif-off commented on data/filedefs/filetypes.autoit:
```diff
@@ -0,0 +1,72 @@
+#settings for folding comment blocks and preprocessor lines
+fold.comment=1
+fold.compact=0
+fold.preprocessor=1
```
I added default values from AutoIt Lexer just in case because I found them in ```LexAU3.cxx``` only. I think I will comment them.

Skif-off replied:
Ok.

Skif-off commented on ctags/parsers/autoit.c:
```diff
@@ -0,0 +1,121 @@
+		{
+			/* min. string "#region" > 7 */
+			if ((p [1] == 'R' || p [1] == 'r') &&
+				strlen ((const char *) p) > 8 &&
```
How can I use subscript (for examle ```p [1]```) if I don't know subscript dimension range, i.e. string length?

b4n replied:
C strings have a last element with value `0`, which is actually delimiting the end of the string. `strlen(str)` is just going through `str` to find at which offset `0` is found (it could basically be implemented as `for (len = 0; str[len] != 0; len++);`).

Anyway, this means that so long as none of the characters before the subscript has value `0`, the next one is a valid subscript -- it just can be the last one if it has value `0`.  Thus, as you check the values of subscripts successively against a defined set *not including* `0`, you're sure that you won't try subscripting past the end of the string, otherwise your tests would have to be false already anyway.
So if `p[1]` is either `'R'` or `'r'` (so not `0`), `p[2]` has to be a valid subscript.

Skif-off replied:
I remembered why I used ```strlen()``` :) Besides subscript.
Function defining:
```autoit
Func t()
```
function name is ```t``` and one symbol is a minimum length of name. If string beginning with ```func``` then we have define of function , but if string length less 8 symbols then we have not valid function name, i.e. we have not function name, and this string should be discarded.

b4n replied:
I still think it's irrelevant, because it's then an "optimization" (as the code after more correctly checks the syntax), but it only kicks in when the there is less than an extra 8 characters after the line starting with `f`. This is highly unlikely to result in any speed benefits (especially as you're checking the next characters for `unc[ \t]`), and is highly likely to let pass things that aren't a function anyway but just happen to be long enough.

In any case, unless profiling shows this is faster (I doubt it), the clearest/simplest thing should be used IMO, which IMO again, is dropping this check.

Skif-off commented on ctags/parsers/autoit.c:
```diff
@@ -0,0 +1,121 @@
+				(p [3] == 'G' || p [3] == 'g') &&
+				(p [4] == 'I' || p [4] == 'i') &&
+				(p [5] == 'O' || p [5] == 'o') &&
+				(p [6] == 'N' || p [6] == 'n'))
```
Valid syntax:
```autoit
#Region
#Region ;foobar something
#Region;foobar something
#Region foobar something
```
In other cases string will be ignored (and of course in first case it will not be shown in the list of symbols too). Works fine, I use it long time ago.

b4n replied:
Hum, I just tested whether I'm wasn't crazy, and I ain't:
```autoit
#regiontest
#endregion
```
results in showing a region named `test` in the symbols tree.

It might be considered OK as this is probably invalid syntax, I'm just mentioning this so you can make an informed decision whether it's a problem or an acceptable behavior given invalid input not worth handling.

---
Additionally with your examples, the leading `;` is included in the displayed region name, is that intentional? If not, you probably could simply use
```c
while (isspace ((int) *p) || *p == ';')
```
below.  It would allow for `#region;;;;foo`, but that might be valid or OK to get wrong (as suggested above, it can be considered OK to emit improper tags in invalid input)

b4n replied:
I'm not sure, but I actually expect the `;` to start a standard comment, so it's probably an unnamed region, with just a comment that happens to be there.

Skif-off replied:
I think yes, most likely, but it's used, it works (in other editors too), it's not a glaring mistake and I decided to do so.

Skif-off commented on ctags/parsers/autoit.c:
```diff
@@ -0,0 +1,121 @@
+/*
```
Sorry, I can not check it in Geany (maybe because CTags version is more old?), I could not transfer the changes to my PR.
I checked your PR in CTags and it seems that working fine.

b4n replied:
Yeah it's using features we don't yet have, but there's a PR on importing that, I just have to finish reviewing it (and it's huge).

Skif-off replied:
Is it your local, not public PR for now (i.e. in your PC)? I could not find it here and in your fork of Geany's repository.

elextr replied:
I think @b4n means #1263, but even that is almost 2 years olde because its so big.

Skif-off replied:
> and it's huge

I don't know how to express the degree (level?) of my impression on English :))

elextr replied:
"Oh [expletive deleted] " :grin:

Skif-off replied:
But it can be try to use right now, isn't it?

Skif-off replied:
I tried to use [#1263](https://github.com/geany/geany/pull/1263): as I understand, [#1263](https://github.com/geany/geany/pull/1263) requires updating from Geany's upstream and from UCTags's upstream too...
I'm sorry that I can't help :(

Skif-off commented:
@b4n can you look at this [parser](https://github.com/Skif-off/geany/blob/30036ffee8062fb9fb466eb485a3f9a82ff03632/ctags/parsers/autoit.c), if you have free time? I maked small update from UCTags's upstream, (until UCTags is updated in Geany).

Skif-off commented:
Hm... Travis is fail with ``` MINGW=yes```, but as I understood it's not my fault, isn't it?

Skif-off commented:
New parser from UCtags doesn't work (Geany fail after start).
Is it problem with some new/latest feature(s) of UCtags?

codebrainz commented:
> Geany fail after start

Can you elaborate? A [backtrace](https://www.geany.org/support/bugs) would help troubleshoot.

Skif-off commented:
I forgot that under Linux instead Win will be more detailed:
```
Starting program: /usr/local/bin/geany -v
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

(process:19323): Tagmanager-ERROR **: 13:32:37.609: Different number of tag types in TM (2) and ctags (5) for AutoIt

Program received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff5563e11 in ?? () from /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
(gdb) bt
#0  0x00007ffff5563e11 in  () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#1  0x00007ffff5564eac in g_log_default_handler () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#2  0x00007ffff7944f3c in handler_log (domain=0x7ffff7b5d8bd "Tagmanager", level=6, msg=0x55555578d2b0 "Different number of tag types in TM (2) and ctags (5) for AutoIt", data=<optimized out>)
    at log.c:136
#3  0x00007ffff556513d in g_logv () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#4  0x00007ffff55652af in g_log () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#5  0x00007ffff7b00861 in tm_parser_verify_type_mappings () at tm_parser.c:698
#6  0x00007ffff7b03827 in tm_create_workspace () at tm_workspace.c:83
#7  0x00007ffff7b03827 in tm_get_workspace () at tm_workspace.c:121
#8  0x00007ffff7945f97 in main_lib (argc=<optimized out>, argv=<optimized out>) at libmain.c:1070
#9  0x00007ffff74f8b97 in __libc_start_main (main=
    0x555555554600 <main>, argc=2, argv=0x7fffffffe0c8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe0b8) at ../csu/libc-start.c:310
#10 0x000055555555463a in _start ()
```

Skif-off commented:
Hmm... It's my fault: I thought that I must write in ```TMParserMapEntry``` only nedeed symbols.

Skif-off commented:
Any news?

Skif-off commented:
[/jobs/6309415770](https://github.com/geany/geany/actions/runs/3720132501/jobs/6309415770): What does

> (process:10108): Tagmanager-WARNING **: 12:59:46.979: Not all tag types mapped to symbol tree groups for AutoIt

mean?
Maybe just drop the test when Meson is used?

elextr commented:
> (process:10108): Tagmanager-WARNING **: 12:59:46.979: Not all tag types mapped to symbol tree groups for AutoIt

https://github.com/geany/geany/blob/bb45e3b7485e27f29bf652968b9094da322d0bc7/src/tagmanager/tm_parser.c#L1351

Tagmanager and ctags are out of sync for autoit, did your updated parser change something?

Skif-off commented:
Parser and lexer are current versions from upstreams (and both have not been updated for a long time). I'm sorry, right now I don't understand what the problem is :)
ctags/parsers/autoit.c:
```c
static roleDefinition AutoItIncludeRoles [] = {
	{ true, "system", "system include" },
	{ true, "local", "local include" },
};

static kindDefinition AutoItKinds [] = {
	{ true, 'f', "func", "functions" },
	{ true, 'r', "region", "regions" },
	{ true, 'g', "global", "global variables" },
	{ true, 'l', "local", "local variables" },
	{ true, 'S', "script", "included scripts",
	  .referenceOnly = true, ATTACH_ROLES (AutoItIncludeRoles) },
};
```

src/tagmanager/tm_parser.c:
```c
static TMParserMapEntry map_AUTOIT[] = {
	{'f', tm_tag_function_t},
	{'r', tm_tag_other_t},
	{'g', tm_tag_member_t},
	{'l', tm_tag_member_t},
	{'S', tm_tag_member_t},
};
static TMParserMapGroup group_AUTOIT[] = {
	{_("Functions"), TM_ICON_METHOD, tm_tag_function_t},
	{_("Regions"), TM_ICON_OTHER, tm_tag_other_t},
};
```

`AutoItKinds` and `map_AUTOIT` are the same size and as I understand and see, `map_AUTOIT` and `group_AUTOIT` do not have to match in size.

elextr commented:
Then something has probably changed in all the big changes in TM recently, @techee?

Skif-off commented:
I took a quick look the list of commits and replaced `tm_tag_member_t` with `tm_tag_undef_t` (I tried to remember why I used `tm_tag_member_t` but couldn't): fail.

techee commented:
> (process:10108): Tagmanager-WARNING **: 12:59:46.979: Not all tag types mapped to symbol tree groups for AutoIt

It means that each `tm_tag_...` from
```
static TMParserMapEntry map_AUTOIT[] = {
	{'f', tm_tag_function_t},
	{'r', tm_tag_other_t},
	{'g', tm_tag_member_t},
	{'l', tm_tag_member_t},
	{'S', tm_tag_member_t},
};
```
has to appear somewhere in
```
static TMParserMapGroup group_AUTOIT[] = {
	{_("Functions"), TM_ICON_METHOD, tm_tag_function_t},
	{_("Regions"), TM_ICON_OTHER, tm_tag_other_t},
};
```
In this case, you are missing the `tm_tag_member_t `. To solve this, you could create a separate root such as
```
	{_("Members"), TM_ICON_OTHER, tm_tag_member_t},
```
or add it to an existing root such as
```
	{_("Functions"), TM_ICON_METHOD, tm_tag_function_t | tm_tag_member_t},
```
(whatever makes sense for this language).

Skif-off commented:
@techee , yes, I figured it out, I found your commits in upstream: I replaced `tm_tag_member_t` with `tm_tag_undef_t` and updated the `simple.au3.tags` file. but again I got an [error](https://github.com/geany/geany/actions/runs/3733542779/jobs/6334467745)  (but no details now).

elextr commented:
@Skif-off [this](https://github.com/geany/geany/commit/ad304cf79afbf4dd8080417abbdd27e904b521a9) gotcha, the tag file format is pretty printed before `diff`ing.

Skif-off commented:
@elextr , interesting, `simple.html.tags`, for example,  contents the same characters, but without error...
I will check manually, maybe just disable testing.

elextr commented:
No `simple.html.tags` was changed [here](https://github.com/geany/geany/commit/aba66f07cca62b28d33534ec88c686b600fe2533#diff-40f1bf07e75d4cf91cce257030fa3b92f7b66a5aae59339cd871f44dc4e8ea32) to match the new format.

elextr commented:
Just process your reference tags file with [this](https://github.com/geany/geany/blob/master/scripts/print-tags.py) so it will match the test file that is also processed that way.

Skif-off commented:
@elextr , @techee , thanks! Now works fine.

techee commented:
> @elextr , @techee , thanks! Now works fine.

Great!

We forgot to update the documentation with the pretty-printing script though - I created #3353. @elextr Does it look OK to you?

elextr commented:
@b4n @techee are you both happy to merge?

techee commented:
Maybe a question - by just having a look at the parser (and not knowing the language), it seems that the parser supports scopes (i.e. nesting one tag inside other tag). The provided unit test only shows some simple script without this. But if scopes are supported, then it should be mentioned inside `tm_parser_has_full_scope()`. Also would be nice to have a unit test that covers scope generation.

I played with the provided unit test a bit and it seems that you can nest one function inside another - in this case the Symbols tree in Geany shows these tags separated by dot instead of displayed as a tree. Adding `TM_PARSER_AUTOIT` to 
`tm_parser_has_full_scope()` to indicate it has full scope info should fix that.

Second, very minor thing is that in https://github.com/geany/geany/pull/3428 we changed comments for `wordchars` and `comment_use_indent` in the filetype config files and they should be changed here too (but it's something we can do once this PR is merged).

Apart from these minor and trivial to fix problems everything seems to be present, syntax gets highlighted, tags are generated so I think it can be merged.

Skif-off commented:
> by just having a look at the parser (and not knowing the language), it seems that the parser supports scopes (i.e. nesting one tag inside other tag).
> ...
> it seems that you can nest one function inside another

It seems the language does not support such constructs.

Any part of code can be placed inside the `#Region ... #EndRegion`, but `#Region ... #EndRegion` is not a regular operator, macro or directive and does not participate in code execution in any way: this is just a visual cue (for the authors, not for the AutoIt), which allows to fold pieces of code. Usually these are some logically related parts of the script - group of functions and/or variables and so on.
I'm not sure if I managed to explain correctly, I'm one of those who just use it :)

> Second, very minor thing is that in https://github.com/geany/geany/pull/3428 we changed comments for wordchars and comment_use_indent in the filetype config files

Fixed.

techee commented:
> It seems the language does not support such constructs.

OK, I mentioned this because I noticed the use of `nestingLevel...()` and `CORK` stuff in the parser which are used for scopes. 

> Any part of code can be placed inside the #Region ... #EndRegion, but #Region ... #EndRegion is not a regular operator, macro or directive and does not participate in code execution in any way: this is just a visual cue (for the authors, not for the AutoIt), which allows to fold pieces of code. Usually these are some logically related parts of the script - group of functions and/or variables and so on.

Yes, but these get parsed and the scope information gets reported for tags so this parser uses scopes. Even though in this case the scope has at most 2 members, I would suggest putting the parser into the group inside `tm_parser_has_full_scope()` that reports `TRUE` - the second group is more or less reserved for parsers that should return full scope but are a bit buggy which this parser isn't. This will help us keep track of which parsers report scopes and which don't - otherwise this change should have no effect on the unit tests or the symbol tree.

>> Second, very minor thing is that in https://github.com/geany/geany/pull/3428 we changed comments for wordchars and comment_use_indent in the filetype config files

> Fixed.

Great, thanks.

techee commented:
> Done.

Thanks. The PR looks good to go from my perspective.

> To be honest, I don't see the difference between TRUE and FALSE :)

You won't in this case because the language itself supports only one level of nesting. The change is more for us so we know which languages support scope generation and which don't.

b4n commented on src/tagmanager/tm_parser.c:
```diff
@@ -1080,6 +1080,18 @@ static TMParserMapGroup group_BATCH[] = {
+static TMParserMapEntry map_AUTOIT[] = {
+	{'f', tm_tag_function_t},
+	{'r', tm_tag_other_t},
+	{'g', tm_tag_undef_t},
+	{'l', tm_tag_undef_t},
```
Don't we want to see at least global variables?  And even local ones would be nice, esp. as IIUC they are now not shown but can still be used for autocompletion (@techee?)

techee replied:
Yes, that would be nice I think (I didn't mention this because I thought they were disabled for some language-specific reason).

If you decide to enable local variables, please map them to `tm_tag_local_var_t` - this mapping is a bit special as it doesn't make the local variables appear in the sidebar (which is a bit too verbose with them) and only shows them in the autocompletion popup.

techee replied:
Plus we want to have all the enabled tags covered by unit tests - I haven't checked if the provided unit test contains local and global tags.

Skif-off replied:
I just wanted to add minimal guaranteed working AutoIt support and don't know all the ins and outs of how Geany works.
And didn't want to bloat the list of symbols (maybe in the future if anyone wants to).

I.e. `tm_tag_variable_t` for global variables and `tm_tag_local_var_t` for local?

techee replied:
I'd suggest that we address this in a separate PR (@b4n what do you think?). This PR works well as it is and I can prepare a PR for the global and local variables afterwards - it should be trivial.

b4n replied:
@techee sure, that'd be fine as well (although it'd probably get forgotten for a decade or two until somebody notices it from scratch again)

b4n reviewed (commented):
Apart from the comment, and assuming the parsers and lexers are upstream's, LGTM (esp. if @techee likes it as well :))

techee commented:
OK, I've just merged this PR so we are finally done with it. @Skif-off thanks for your contribution and for your patience!

> @techee sure, that'd be fine as well (although it'd probably get forgotten for a decade or two until somebody notices it from scratch again)

@b4n I'm going to do it right now so we don't forget.
